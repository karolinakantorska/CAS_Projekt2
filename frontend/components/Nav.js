import React, { useState, useEffect } from "react";
import Link from 'next/link';
import styled from 'styled-components';
import Signout from './Signout';
import User from './User';


const Nav = (props) => {

return (
  <StyledNav>
    <User>
      {(currentUserPermission, currentUserName) => (
        <React.Fragment>
          <p>User: {currentUserName}</p>
          <Link href="/">
            <a>Home</a>
          </Link>
          <Link href="/guides">
            <a>MTB Guides</a>
          </Link>
          {props.currentUserPermission === "ADMIN" ? (
            <Link href="/add_guide">
              <a>Add New MTB Guide</a>
            </Link>
          ) : null}
          {props.currentUserName ? <Signout /> : null}
          {!props.currentUserName ? (
            <Link href="/signup">
              <a>Signup|Signin</a>
            </Link>
          ) : null}
        </React.Fragment>
      )}
    </User>
  </StyledNav>
);
}
const StyledNav = styled.nav`
    background: white;
    display: grid;
    grid-template-columns: repeat(5,1fr) ;
`;

export default Nav;
