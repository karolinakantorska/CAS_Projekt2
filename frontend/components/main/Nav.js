import React from 'react';
import Link from 'next/link';
import styled from 'styled-components';
import Signout from '../signin_signout/Signout';
import User from './User';

const Nav = (props) => {
  return (
    <User>
      {(currentUserPermission, currentUserName) => (
        <StyledNav>
          <div>
            <p>User: {currentUserName}</p>
            <Link href="/">
              <a className="home ">Home</a>
            </Link>
            <Link href="/guides">
              <a className="guides">MTB Guides</a>
            </Link>
            {currentUserPermission === 'ADMIN' && (
              <Link href="/add_guide">
                <a className="add">Add New MTB Guide</a>
              </Link>
            )}
            {currentUserName && (
              <span className="signin">
                <Signout />
              </span>
            )}
            {!currentUserName && (
              <Link href="/signin_page">
                <a className="signin">Signup|Signin</a>
              </Link>
            )}
          </div>
        </StyledNav>
      )}
    </User>
  );
};
const StyledNav = styled.nav`
  background-color: ${(props) => props.theme.color.primary};
  color: ${(props) => props.theme.colorText.negativ};
  div {
    margin: auto;
    max-width: ${(props) => props.theme.maxWidth};
    height: 2.5rem;
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    grid-template-areas: ' user add home guides signin';
  }
  a,
  span {
    color: ${(props) => props.theme.colorText.negativ};
    justify-self: end;
    align-self: center;
  }
  a:hover {
    color: #f5f5f5;
    text-shadow: 0px 0px 40px #ffffff;
  }
  a:active {
  }
  .user {
    grid-area: user;
  }
  .home {
    grid-area: home;
  }
  .add {
    grid-area: add;
  }
  .guides {
    grid-area: guides;
  }
  .signin {
    grid-area: signin;
  }
`;

export default Nav;
