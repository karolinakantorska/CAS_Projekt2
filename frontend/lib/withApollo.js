import withApollo from 'next-with-apollo';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { endpoint } from '../config';


export default withApollo(({ headers }) => {
  return new ApolloClient({
    uri: "http://localhost:4000",
    //        cache: new InMemoryCache().restore(initialState|| {})
    request: (operation) => {
      operation.setContext({
        fetchOptions: {
          credentials: "include",
        },
        headers,
      });
    },
  });
});

/*
function createClient ({ headers }) {
    return new ApolloClient({
        uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
    });
}

export default withApollo(createClient);
*/