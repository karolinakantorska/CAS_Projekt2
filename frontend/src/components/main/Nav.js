import React, { useState } from 'react';
import Link from 'next/link';
import styled from 'styled-components';

import { Icon } from '@rmwc/icon';
//Components
import Loading from '../reusable/LoadingBar';
import ErrorGraphql from '../reusable/ErrorGraphql';
import Signout from '../signin_signout/Signout';
import NavAdmin from '../main/NavAdmin';

import { useCurrentUser } from '../../apollo/querries/useCurrentUser';
import { StyledTextMenuBlack, StyledTextBody2 } from '../styles/StyledText';

const Nav = () => {
  const [show, setShow] = useState(true);
  const { loading, error, data } = useCurrentUser();

  if (loading) {
    return <Loading />;
  }
  if (error) {
    return <ErrorGraphql error={error} />;
  }
  if (data) {
    return (
      <>
        <StyledNav>
          <StyledTextBody2 className="user">
            {data.currentUser.name ? data.currentUser.name : 'please login'}
          </StyledTextBody2>
          <Link href="/">
            <StyledTextMenuBlack className="home">Home</StyledTextMenuBlack>
          </Link>
          <StyledTextMenuBlack
            className={`userSpec ${show && 'active'}`}
            onClick={() => setShow((show) => setShow(!show))}
          >
            Info
          </StyledTextMenuBlack>
          <Link href="/guides">
            <StyledTextMenuBlack className="guides">MTB Guides</StyledTextMenuBlack>
          </Link>
          {data.currentUser.permissions === 'ADMIN' && (
            <StyledTextMenuBlack
              className={`userSpec ${show && 'active'}`}
              onClick={() => setShow((show) => setShow(!show))}
            >
              Add Guide
            </StyledTextMenuBlack>
          )}
          {data.currentUser.name && (
            <span className="signin">
              <Signout />
            </span>
          )}
          {!data.currentUser.name && (
            <Link href="/signin_page">
              <StyledTextMenuBlack className="signin">
                <Icon icon="person_outline" aria-label="Login" />
              </StyledTextMenuBlack>
            </Link>
          )}
        </StyledNav>
        <StyledSpan>{show && <NavAdmin />}</StyledSpan>
      </>
    );
  }
};

const StyledNav = styled.nav`
  cursor: pointer;
  margin: auto;
  //max-width: var(--maxWidth);
  display: grid;
  justify-items: center;
  align-items: center;
  grid-template-rows: 2fr 4fr;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-areas:
    ' none none none user'
    ' home userSpec guides signin';
  white-space: nowrap;
  background-color: rgba(255, 255, 255, 1);
  a:hover,
  .active {
    color: var(--colorSecundary);
  }
  .user {
    cursor: auto;
    grid-area: user;
  }
  .home {
    grid-area: home;
  }
  .userSpec {
    grid-area: userSpec;
  }
  .guides {
    grid-area: guides;
  }
  .signin {
    grid-area: signin;
  }
`;
const StyledSpan = styled.span`
  display: grid;
  grid-template-rows: 30px;
`;
export default Nav;
