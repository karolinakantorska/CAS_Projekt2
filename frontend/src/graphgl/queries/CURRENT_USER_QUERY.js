import { gql } from '@apollo/client';

const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    currentUser {
      id
      email
      name
      permissions
    }
  }
`;

export default CURRENT_USER_QUERY;
