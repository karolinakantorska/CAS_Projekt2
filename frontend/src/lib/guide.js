const specialisation = [
  'Trail riding',
  'Singletrack',
  'Bikepark',
  'Downhill',
  'Cross country',
  'All mountain',
  'Enduro',
  'Dirt jumping',
  'Freeride',
  'MTB touring',
  'Fatbiking',
  'Ebikes',
];

export const guideAdditionalInfo = {
  // reservations should not be obligatory, no !
  specialisations: specialisation,
  ebike: true,
  mtb: true,
  title: 'The Best Guide in The Valley',
};
