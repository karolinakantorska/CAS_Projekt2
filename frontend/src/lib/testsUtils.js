export const fakeReservation = () => {
  return {
    __typename: 'Reservation',
    description: '',
    guide: {
      __typename: 'User',
      id: 'ckiemg0xj9qx70981qd281li5',
      name: 'Olaf',
      photo:
        'https://res.cloudinary.com/karolinauploads/image/upload/v1607349564/images/asxyln6ja3ahgfej1bci.jpg',
      surname: 'Olaf',
    },
    relatedDay: {
      __typename: 'Day',
      day: '19',
      month: 'December',
      year: '2020',
    },
    time: 'AM',
    userEmail: 'karolina@gmail.com',
    userName: 'karolina',
  };
};
export const fakeGuide = {
  __typename: 'User',
  description: 'Hallo',
  email: 'olaf@gmail.com',
  id: '123',
  name: 'Olaf',
  photo:
    'https://res.cloudinary.com/karolinauploads/image/upload/v1607349564/images/asxyln6ja3ahgfej1bci.jpg',
  surname: 'Olafski',
  permissions: 'GUIDE',
};
export const fakeUser = {
  __typename: 'User',
  email: 'magic@gmail.com',
  id: '123',
  name: 'Magic',
  surname: 'Michael',
  permissions: 'USER',
};
export const fakeAdmin = {
  __typename: 'User',
  email: 'karolina@gmail.com',
  id: '123',
  name: 'Karolina',
  permissions: 'ADMIN',
};
export const fakeGuides = [
  {
    __typename: 'User',
    description: '',
    email: 'speady@gmail.com',
    id: 'ckh3n7algb8r50946cpug61su',
    name: 'Speedy',
    photo:
      'https://res.cloudinary.com/karolinauploads/image/upload/v1604508914/images/g65d0lylojpsrqqqzpgu.jpg',
    surname: 'Gonzales',
  },
  {
    __typename: 'User',
    description: '',
    email: '<fdgt',
    id: 'ckiemg0xj9qx70981qd281li5',
    name: 'Olaf',
    photo:
      'https://res.cloudinary.com/karolinauploads/image/upload/v1607349564/images/asxyln6ja3ahgfej1bci.jpg',
    surname: 'Olaf',
  },
];
//days:
export const fakeMonthReservations = [
  {
    __typename: 'Day',
    day: '',
    month: 'December',
    year: '2020',
    reservations: [
      {
        __typename: 'Reservation',
        id: 'cki61cvx81i560952hc0jcwfh',
        time: 'AM',
        userEmail: 'karol@gmail.com',
        userName: 'karol',
      },
    ],
  },
  {
    __typename: 'Day',
    day: '18',
    month: 'December',
    year: '2020',
    reservations: [
      {
        __typename: 'Reservation',
        id: 'ckipt7r5a4q8l098109ma3e42',
        time: 'AM',
        userEmail: 'magda@gmail.com',
        userName: 'magda',
      },
      {
        __typename: 'Reservation',
        id: 'ckipvr76876jp09956jm1opk4',
        time: 'PM',
        userEmail: 'magda@gmail.com',
        userName: 'magda',
      },
    ],
  },
];
