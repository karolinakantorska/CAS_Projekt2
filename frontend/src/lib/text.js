export const text = `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Neque sodales ut etiam sit amet nisl purus in. At imperdiet dui accumsan sit amet nulla facilisi. Tellus at urna condimentum mattis pellentesque id nibh tortor id. Posuere ac ut consequat semper viverra nam. Magna fringilla urna porttitor rhoncus dolor purus. Nascetur ridiculus mus mauris vitae ultricies. Eros in cursus turpis massa tincidunt. Parturient montes nascetur ridiculus mus. Sit amet commodo nulla facilisi nullam vehicula ipsum. Ut morbi tincidunt augue interdum. Enim eu turpis egestas pretium aenean pharetra. Cursus risus at ultrices mi tempus imperdiet.`;
