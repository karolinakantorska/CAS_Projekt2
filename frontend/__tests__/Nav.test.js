import { MockedProvider } from '@apollo/client/testing';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import Nav from '../src/components/main/Nav';
import CURRENT_USER_QUERY from '../src/graphgl/queries/CURRENT_USER_QUERY';
import SIGNIN_MUTATION from '../src/graphgl/mutations/SIGNIN_MUTATION';
import { fakeUser, fakeAdmin } from '../src/lib/testsUtils';

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { currentUser: null } },
  },
];
const signedInMocksUser = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { currentUser: fakeUser } },
  },
];
const signedInMocksAdmin = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { currentUser: fakeAdmin } },
  },
];

describe('<Nav />', () => {
  xit('renders proper <Nav /> when signed out', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    const nav = wrapper.find('[data-testid="nav"]');
    //console.log(nav.debug());
    //expect(toJSON(nav)).toMatchSnapshot();
    const navChildren = nav.find('a');
    expect(navChildren.children().length).toBe(4);
    expect(wrapper.containsMatchingElement(<a>User:</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>Home</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>MTB Guides</a>)).toBeTruthy();
  });
  xit('renders proper <Nav /> with user name when signed in as User', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksUser}>
        <Nav />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    const nav = wrapper.find('[data-testid="nav"]');
    //console.log(nav.debug());
    expect(toJSON(nav)).toMatchSnapshot();
    const navChildren = nav.find('a');
    expect(navChildren.children().length).toBe(5);
    expect(wrapper.containsMatchingElement(<a>User: Magic</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>Home</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>MTB Guides</a>)).toBeTruthy();
  });

  xit('renders proper <Nav /> when signed in as Admin', async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksAdmin}>
        <Nav />
      </MockedProvider>,
    );
    await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    const nav = wrapper.find('[data-testid="nav"]');
    //console.log(nav.debug());
    expect(toJSON(nav)).toMatchSnapshot();
    const navChildren = nav.find('a');
    expect(navChildren.children().length).toBe(6);
    expect(wrapper.containsMatchingElement(<a>User: {fakeAdmin.name}</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>Add MTB Guide</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>Home</a>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<a>MTB Guides</a>)).toBeTruthy();
  });
});
