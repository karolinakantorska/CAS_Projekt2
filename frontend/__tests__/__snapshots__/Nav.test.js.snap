// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders proper <Nav /> when signed in as Admin 1`] = `
<div
  data-testid="nav"
>
  <StyledText__StyledTextMenuWhite
    className="user"
  >
    <a
      className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 user"
    >
      User: 
      Karolina
    </a>
  </StyledText__StyledTextMenuWhite>
  <Link
    href="/"
  >
    <StyledText__StyledTextMenuWhite
      className="home "
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 home "
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Home
      </a>
    </StyledText__StyledTextMenuWhite>
  </Link>
  <Link
    href="/guides"
  >
    <StyledText__StyledTextMenuWhite
      className="guides"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 guides"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        MTB Guides
      </a>
    </StyledText__StyledTextMenuWhite>
  </Link>
  <Link
    href="/add_guide"
  >
    <StyledText__StyledTextMenuWhite
      className="add"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 add"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Add MTB Guide
      </a>
    </StyledText__StyledTextMenuWhite>
  </Link>
  <span
    className="signin"
  >
    <Signout>
      <StyledText__StyledTextMenuWhite
        onClick={[Function]}
      >
        <a
          className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6"
          onClick={[Function]}
        >
          <Icon
            aria-label="Logout"
            icon="person"
          >
            <ForwardRef(IconRoot)
              aria-label="Logout"
              className="rmwc-icon rmwc-icon--ligature material-icons"
            >
              <ForwardRef(Tag)
                aria-label="Logout"
                className="rmwc-icon rmwc-icon--ligature material-icons"
                tag="i"
              >
                <i
                  aria-label="Logout"
                  className="rmwc-icon rmwc-icon--ligature material-icons"
                >
                  person
                </i>
              </ForwardRef(Tag)>
            </ForwardRef(IconRoot)>
          </Icon>
        </a>
      </StyledText__StyledTextMenuWhite>
    </Signout>
  </span>
</div>
`;

exports[`<Nav /> renders proper <Nav /> with user name when signed in as User 1`] = `
<div
  data-testid="nav"
>
  <StyledText__StyledTextMenuWhite
    className="user"
  >
    <a
      className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 user"
    >
      User: 
      Magic
    </a>
  </StyledText__StyledTextMenuWhite>
  <Link
    href="/"
  >
    <StyledText__StyledTextMenuWhite
      className="home "
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 home "
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Home
      </a>
    </StyledText__StyledTextMenuWhite>
  </Link>
  <Link
    href="/guides"
  >
    <StyledText__StyledTextMenuWhite
      className="guides"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6 guides"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        MTB Guides
      </a>
    </StyledText__StyledTextMenuWhite>
  </Link>
  <span
    className="signin"
  >
    <Signout>
      <StyledText__StyledTextMenuWhite
        onClick={[Function]}
      >
        <a
          className="StyledText__StyledTextMenuWhite-sc-1yfchkt-6"
          onClick={[Function]}
        >
          <Icon
            aria-label="Logout"
            icon="person"
          >
            <ForwardRef(IconRoot)
              aria-label="Logout"
              className="rmwc-icon rmwc-icon--ligature material-icons"
            >
              <ForwardRef(Tag)
                aria-label="Logout"
                className="rmwc-icon rmwc-icon--ligature material-icons"
                tag="i"
              >
                <i
                  aria-label="Logout"
                  className="rmwc-icon rmwc-icon--ligature material-icons"
                >
                  person
                </i>
              </ForwardRef(Tag)>
            </ForwardRef(IconRoot)>
          </Icon>
        </a>
      </StyledText__StyledTextMenuWhite>
    </Signout>
  </span>
</div>
`;
